### PR TITLE
adding last modification tag

### PR DESCRIPTION
**Pull request by eishay**
@raymondng @andrewconner

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 1]
```diff
@@ -436,4 +437,9 @@ class KeepRepoImpl @Inject() (
     import StaticQuery.interpolation
     sql"""select b.library_id, count(*) as cnt from bookmark b, library l where l.id = b.library_id and l.state='active' and l.visibility='published' and b.created_at > $since group by b.library_id order by count(*) desc limit $count""".as[(Id[Library], Int)].list
   }
+
+  def latestKeepInLibrary(libraryId: Id[Library])(implicit session: RSession): Option[DateTime] = {
```
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:441]: :+1: Was thinking the same. We should probably cache something like this as well.

File: kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala [hunk 2]
```diff
@@ -73,27 +77,39 @@ class SiteMapGenerator @Inject() (
...
-    val paths = libraries.filter { lib =>
+    val libs = libraries.filter { lib =>
       owners.get(lib.ownerId).isDefined && db.readOnlyMaster { implicit ro => keepRepo.getCountByLibrary(lib.id.get) > 3 } // proxy for quality; need bulk version
-    } map { lib =>
+    }
+
+    def path(lib: Library): String = {
       val owner = owners(lib.ownerId)
-      s"https://www.kifi.com${Library.formatLibraryPath(owner.username, owner.externalId, lib.slug)}"
+      s"${fortyTwoConfig.applicationBaseUrl}{Library.formatLibraryPath(owner.username, owner.externalId, lib.slug)}"
```
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala:86]: Thanks!

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala (hunk 1 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:441]: Probably should name it something like `lastUpdatedAt` for libraries.

File: kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala [hunk 3]
```diff
@@ -73,27 +77,39 @@ class SiteMapGenerator @Inject() (
...
+    }
+
+    def lastMod(lib: Library): DateTime = {
+      db.readOnlyReplica { implicit s =>
+        keepRepo.latestKeepInLibrary(lib.id.get) match {
+          case Some(date) if date.isAfter(lib.updatedAt) => date
+          case _ => lib.updatedAt
+        }
+      }
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala:95]: This won’t update the last-modified date if someone removes a keep from a library, which could sometimes be important. Probably worth explicitly storing a `last_modified_at` field in each library—if not today, then eventually. We’ve been bitten by basing logic on `createdAt` and `updatedAt` in the past.

Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/common/seo/SiteMapGenerator.scala:95]: agree

**Pull request merged**